environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/NetworkCreationHandler.java

```diff
@@ -114,6 +126,50 @@ private void setChildEnvironmentNetworkIfItHasParentWithTheSameCloudProvider(Env
         }
     }
 
+    private Map<String, CloudSubnet> removePrivateSubnets(Map<String, CloudSubnet> endpointGatewaySubnetMetas) {
+        LOGGER.debug("Removing any private subnets from the provided endpoint gateway list because they won't be used.");
+        if (endpointGatewaySubnetMetas == null || endpointGatewaySubnetMetas.isEmpty()) {
+            return Map.of();
+        }
+        return endpointGatewaySubnetMetas.entrySet().stream()
+            .filter(entry -> !entry.getValue().isPrivateSubnet())
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private void validateSubnetsIfProvided(Environment environment, Map<String, CloudSubnet> subnetMetas,
+            Map<String, CloudSubnet> endpointGatewaySubnetMetas) {
+        if (hasNetwork(environment) && environment.getNetwork().getPublicEndpointAccessGateway() == PublicEndpointAccessGateway.ENABLED &&
+                environment.getNetwork().getRegistrationType() != RegistrationType.CREATE_NEW) {
+            if (subnetMetas != null && !subnetMetas.isEmpty()) {
+                if (endpointGatewaySubnetMetas != null && !endpointGatewaySubnetMetas.isEmpty()) {
+                    LOGGER.debug("Verifying that provided endpoint gateway subnets share availability zones with provided environment subnets.");
+                    Set<String> subnetAZs = subnetMetas.values().stream()
+                        .map(CloudSubnet::getAvailabilityZone)
+                        .collect(Collectors.toSet());
+                    Set<String> endpointGatewaySubnetAZs = endpointGatewaySubnetMetas.values().stream()
+                        .map(CloudSubnet::getAvailabilityZone)
+                        .collect(Collectors.toSet());
+                    if (!subnetAZs.equals(endpointGatewaySubnetAZs)) {
```

Comment:
Should this check be a subset check rather than equals() check? We can have env with 3 subnets, but endpoint gateway with 1 subnet, no?

Reply:
No, we can't. The load balancer needs subnets in the same AZs as all the EC2 instances it's routing to. So if our private subnets are in az-1, az-2, and az-3, and our public subnets are in az-1 and az-2, then if an EC2 instance is created in az-3 it's completely inaccessible to the load balancer. In order to prevent that possibility, we require the users to provide both private and public subnets in all the same AZs, which is what this check is doing here.

Reply:
My understanding is that the initial set of subnets given at the time of the env creation can contain more subnets than actually gets used for DL/DH instances (like subnets for dwx, liftie, etc). In DH you can select which subnet to use, but apparently not for DL? 

Maybe this is fine, as long as somebody from CB who knows the subnet selection logic takes a look at it.

Reply:
That's true. Not every subnet in the lists being processed here will be used. Which is really why I have this logic. At this stage I don't know which of the subnets will be selected, so to make sure there won't be any hiccups in the eventual subnet selection logic I just make sure that we have subnets with matching AZs here and fail immediately if not.